stock_dynamic_routing/models/stock_move.py

```diff
@@ -0,0 +1,443 @@
+        """Apply pull dynamic routing
+
+        When a move has a dynamic routing configured on its location and the
+        destination of the move does not match the destination of the routing,
+        this method updates the move's destination and it's picking type with
+        the routing ones and creates a new chained move after it.
+        """
+        pickings_to_check_for_emptiness = self.env["stock.picking"]
+        for move in self:
+            routing_rule = routing_details[move].rule
+            if not routing_rule.method == "pull":
+                continue
+
+            if move.picking_id.picking_type_id == routing_rule.picking_type_id:
+                # already correct
+                continue
+
+            # we expect all the lines to go to the same destination for
+            # pull routing rules
+            original_destination = move.location_dest_id
+            current_picking_type = move.picking_id.picking_type_id
+            move.with_context(
+                __applying_routing_rule=True
+            ).location_id = routing_rule.location_src_id
+            move.picking_type_id = routing_rule.picking_type_id
+            dest_location = move.location_dest_id
+            rule_location = routing_rule.location_dest_id
+            if rule_location.is_sublocation_of(dest_location):
+                # The destination of the move, as a parent of the destination
+                # of the routing, goes to the correct place, but is not precise
+                # enough: set the new destination to match the rule's one.
+                # The source of the dest. move will be changed to match it,
+                # which may reapply a new routing rule on the dest. move.
+                move._routing_pull_switch_destination(routing_rule)
+
+            elif not dest_location.is_sublocation_of(rule_location):
+                # The destination of the move is unrelated (nor identical, nor
+                # a parent or a child) to the routing destination: in this case
+                # we have to add a routing move before the current move to
+                # route the goods in the correct place
+                move._routing_pull_insert_move(
+                    routing_rule, current_picking_type, original_destination
+                )
+
+            pickings_to_check_for_emptiness |= move.picking_id
+            move._assign_picking()
+            # Note: we have to call _action_assign() here because if the move
+            # has been split because of partial availability, we want to ensure
+            # to reserve the move which has been "routed" first. Even if
+            # _action_assign() is called again, it should not be an issue
+            # because the move's state will be "assigned" and will be excluded
+            # by the method.
+            move.with_context(exclude_apply_dynamic_routing=True)._action_assign()
+
+        pickings_to_check_for_emptiness._dynamic_routing_handle_empty()
```

Comment:
If I understand correctly, the routed moves are assigned 3 times.
The first, which is rollbacked, to know if yes or no a routing needs to be applied.
A second, right after the routing is applied, to ensure the reserved products are the same as during the 1st action_assign call.
And finally a third in the original `_action_assign` call.
To avoid this 3rd one, we could here return the routed moves.
Then in `_split_and_apply_routing` we could return something like `moves - routed_moves`.
Then, only the moves, which already had their routing applied will be returned to `_action_assign'.
So at the end, `_action_assign` will have been called : 
once for moves with no routing rules.
twice (one roll backed) on moves that already had been routed before
twice on the moves that just had been routed. (instead of 3 times).

What do you think?

Else it seems all good to me

Reply:
The method `_action_assign` ignores moves with a state `assigned` and we know the routed moves have this state (they would not be routed if not assigned, and they can't be partially available because they are split to be available).